io/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+#![no_std]
+
+use gmeta::{In, InOut, Metadata};
+use gstd::{prelude::*, ActorId};
+
+pub type TransactionId = u64;
+
+pub struct StakingMetadata;
+
+impl Metadata for StakingMetadata {
+    type Init = In<InitStaking>;
+    type Handle = InOut<StakingAction, Result<StakingEvent, Error>>;
+    type Others = ();
+    type Reply = ();
+    type Signal = ();
+    type State = IoStaking;
+}
+
+#[derive(Debug, Clone, Decode, Encode, TypeInfo, PartialEq, Eq)]
+pub struct InitStaking {
+    pub staking_token_address: ActorId,
+    pub reward_token_address: ActorId,
+    pub distribution_time: u64,
+    pub reward_total: u128,
+}
+
+#[derive(Debug, Default, Encode, Decode, TypeInfo, Clone, PartialEq)]
+pub struct Staker {
+    pub balance: u128,
+    pub reward_allowed: u128,
+    pub reward_debt: u128,
+    pub distributed: u128,
+}
+
+#[derive(Debug, Clone, Decode, Encode, TypeInfo, PartialEq, Eq)]
+pub enum StakingAction {
+    Stake(u128),
+    Withdraw(u128),
+    UpdateStaking(InitStaking),
+    GetReward,
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub enum StakingEvent {
+    StakeAccepted(u128),
+    Withdrawn(u128),
+    Updated,
+    Reward(u128),
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo, PartialEq)]
+pub enum StakingState {
+    GetStakers,
+    GetStaker(ActorId),
+}
+
+#[derive(Debug, TypeInfo, Encode, Decode, PartialEq)]
+pub enum StakingStateReply {
+    Stakers(Vec<(ActorId, Staker)>),
+    Staker(Staker),
+}
+
+#[derive(Debug, Clone, Default, Encode, Decode, TypeInfo)]
+pub struct IoStaking {
+    pub owner: ActorId,
+    pub staking_token_address: ActorId,
+    pub reward_token_address: ActorId,
+    pub tokens_per_stake: u128,
+    pub total_staked: u128,
+    pub distribution_time: u64,
+    pub produced_time: u64,
+    pub reward_total: u128,
+    pub all_produced: u128,
+    pub reward_produced: u128,
+    pub stakers: Vec<(ActorId, Staker)>,
+    pub transactions: BTreeMap<ActorId, Transaction<StakingAction>>,
+    pub current_tid: TransactionId,
+}
+
+#[derive(Debug, Clone, Encode, Decode, TypeInfo)]
+pub enum Error {
+    NullAmount,
+    NullReward,
+    NullTime,
```

Comment:
```suggestion
    ZeroAmount,
    ZeroReward,
    ZeroTime,
```
I'd use the `Zero*` naming instead of `Null*` because null means more an empty value than zero.